### PR TITLE
Allow transactional metadata update with index creation

### PR DESCRIPTION
**Pull request by danhermann**
Allows clients to supply a function to mutate cluster metadata in the same cluster state update operation that creates an index. The immediate motivation for this is to permit the rollover API to create a new backing index for a data stream and add that index to the data stream in the same cluster state update operation thus simplifying validation and ensuring consistency.

Relates to #53100

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 1]
```diff
@@ -685,10 +701,14 @@ static int getIndexNumberOfRoutingShards(Settings indexSettings, @Nullable Index
...
-                                                BiFunction<ClusterState, String, ClusterState> rerouteRoutingTable) {
-        Metadata newMetadata = Metadata.builder(currentState.metadata())
-            .put(indexMetadata, false)
-            .build();
+                                                BiFunction<ClusterState, String, ClusterState> rerouteRoutingTable,
+                                                BiFunction<Metadata.Builder, IndexMetadata, Metadata.Builder> metadataTransaction) {
+        Metadata.Builder builder = Metadata.builder(currentState.metadata())
+            .put(indexMetadata, false);
+        if (metadataTransaction != null) {
+            builder = metadataTransaction.apply(builder, indexMetadata);
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:709]: Drive-by comment, but does this really need to return a brand new builder? Since the existing builder is being passed in couldn't this be a `BiConsumer<Metadata.Builder, IndexMetadata>` and modify the builder directly?

Comment (danhermann) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:709]: That makes sense to me if the return value on builder methods is only for enabling a fluent API and not for changing the identity of the builder.

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java (hunk 1 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:709]: This PR didn't have any uses of the actual transaction builder, what is the expected use that will change the identity of the builder?

Comment (danhermann) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:709]: @dakrone, it's for rolling over data streams. Create the new backing index and add it to the data stream in the same cluster state update operation.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:709]: I agree that there is no need need for the `metadataTransaction` to return a new builder instance.
Returning the same instance should suffice for the use case we would like to do with this change.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 2]
```diff
@@ -685,10 +701,14 @@ static int getIndexNumberOfRoutingShards(Settings indexSettings, @Nullable Index
      * table based on the live nodes.
      */
     static ClusterState clusterStateCreateIndex(ClusterState currentState, Set<ClusterBlock> clusterBlocks, IndexMetadata indexMetadata,
-                                                BiFunction<ClusterState, String, ClusterState> rerouteRoutingTable) {
-        Metadata newMetadata = Metadata.builder(currentState.metadata())
-            .put(indexMetadata, false)
-            .build();
+                                                BiFunction<ClusterState, String, ClusterState> rerouteRoutingTable,
+                                                BiFunction<Metadata.Builder, IndexMetadata, Metadata.Builder> metadataTransaction) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:705]: Maybe rename `metadataTransaction` to `metadataTransformer`?
Since if provided the function will change the metadata builder.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 3]
```diff
@@ -492,7 +508,7 @@ private ClusterState applyCreateIndexRequestWithExistingMetadata(final ClusterSt
                 // the context is only used for validation so it's fine to pass fake values for the
                 // shard id and the current timestamp
                 indexService.newQueryShardContext(0, null, () -> 0L, null)),
-            Collections.emptyList());
+            Collections.emptyList(), metadataTransaction);
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:511]: nit: `List.of()`?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:709]: maybe in that spirit also add an assertion that the same metadata builder instance is returned from the function?

**Review by martijnvg — COMMENTED** [review]
Looks good, I left some minor nits and I'm agreeing with @dakrone's drive by comment.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java (hunk 1 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:709]: I think we should just turn it into a `BiConsumer` instead?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java (hunk 2 above)
Comment (danhermann) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:705]: Ah, that is a better name 👍

**Review by martijnvg — APPROVED** [review]
LGTM
